app/src/ui/multi-commit-operation/warn-force-push-dialog.tsx

```diff
@@ -0,0 +1,101 @@
+import * as React from 'react'
+import { Checkbox, CheckboxValue } from '../lib/checkbox'
+import { Dispatcher } from '../dispatcher'
+import { DialogFooter, DialogContent, Dialog } from '../dialog'
+import { OkCancelButtonGroup } from '../dialog/ok-cancel-button-group'
+
+interface IWarnForcePushProps {
+  /**
+   * This is expected to be capitalized for correct output on windows and macOs.
+   *
+   * Examples:
+   *  - Rebase
+   *  - Squash
+   *  - Reorder
+   */
+  readonly operation: string
+  readonly dispatcher: Dispatcher
+  readonly askForConfirmationOnForcePush: boolean
+  readonly onBegin: () => void
+  readonly onDismissed: () => void
+}
+
+interface IWarnForcePushState {
+  readonly askForConfirmationOnForcePush: boolean
+}
+
+export class WarnForcePushDialog extends React.Component<
+  IWarnForcePushProps,
+  IWarnForcePushState
+> {
+  public constructor(props: IWarnForcePushProps) {
+    super(props)
+
+    this.state = {
+      askForConfirmationOnForcePush: props.askForConfirmationOnForcePush,
+    }
+  }
+
+  public render() {
+    const { operation, onDismissed } = this.props
+
+    const title = __DARWIN__
+      ? `${operation} Will Require Force Push`
+      : `${operation} will require force push'`
+
+    return (
+      <Dialog
+        title={title}
+        onDismissed={onDismissed}
+        onSubmit={this.onBegin}
+        dismissable={false}
+        type="warning"
+      >
+        <DialogContent>
+          <p>Are you sure you want to {operation}?</p>
+          <p>
+            At the end of the {operation.toLowerCase()} flow, GitHub Desktop
+            will enable you to force push the branch to update the upstream
+            branch. Force pushing will alter the history on the remote and
+            potentially cause problems for others collaborating on this branch.
+          </p>
+          <div>
+            <Checkbox
+              label="Do not show this message again"
+              value={
+                this.state.askForConfirmationOnForcePush
+                  ? CheckboxValue.Off
+                  : CheckboxValue.On
+              }
+              onChange={this.onAskForConfirmationOnForcePushChanged}
+            />
+          </div>
+        </DialogContent>
+        <DialogFooter>
+          <OkCancelButtonGroup
+            okButtonText={
+              __DARWIN__ ? `Begin ${operation}` : `Begin ${operation}`
```

Comment:
It should be lowercase on Windows, right?
```suggestion
              __DARWIN__ ? `Begin ${operation}` : `Begin ${operation.toLowerCase()}`
```

Reply:
Technically you could also do this, which removes some redundancy with the `Begin ` prefix:
```suggestion
              `Begin ${__DARWIN__ ? operation : operation.toLowerCase()}`
```